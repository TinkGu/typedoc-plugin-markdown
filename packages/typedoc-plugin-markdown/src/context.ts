import {
  Application,
  DeclarationReflection,
  ProjectReflection,
} from 'typedoc/dist';
import { PageEvent, RendererEvent } from 'typedoc/dist/lib/output/events';
import { AbstractComponent, BindOption } from 'typedoc/dist/lib/utils';
import { MarkdownRenderer } from './renderer';

class ContextComponent extends AbstractComponent<MarkdownRenderer> {
  @BindOption('entryPoints')
  entryPoints!: string[];
  @BindOption('readme')
  readme!: string;
  @BindOption('entryDocument')
  entryDocument!: string;
  @BindOption('publicPath')
  publicPath!: string;
  @BindOption('hideBreadcrumbs')
  hideBreadcrumbs!: string;
  @BindOption('hidePageTitle')
  hidePageTitle!: string;
  @BindOption('hideInPageTOC')
  hideInPageTOC!: string;
  @BindOption('namedAnchors')
  namedAnchors!: string;
  @BindOption('indexTitle')
  indexTitle!: string;
  @BindOption('filenameSeparator')
  filenameSeparator!: string;
  @BindOption('allReflectionsHaveOwnDocument')
  allReflectionsHaveOwnDocument!: string;

  globalsFile = 'modules.md';

  globalsName =
    this.entryPoints && this.entryPoints.length > 1 ? 'Modules' : 'Exports';

  project: ProjectReflection;

  activeReflection?: DeclarationReflection;

  activeUrl: string;

  initialize() {
    this.listenTo(this.owner, {
      [RendererEvent.BEGIN]: this.onBeginRenderer,
      [PageEvent.BEGIN]: this.onBeginPage,
    });
  }

  onBeginRenderer(event: RendererEvent) {
    this.project = event.project;
  }

  onBeginPage(page: PageEvent) {
    this.activeUrl = page.url;
    this.activeReflection =
      page.model instanceof DeclarationReflection ? page.model : undefined;
  }
}

let context: ContextComponent;

export function loadContext(app: Application) {
  context = new ContextComponent(app.renderer);
}

export function getContext() {
  return context;
}
