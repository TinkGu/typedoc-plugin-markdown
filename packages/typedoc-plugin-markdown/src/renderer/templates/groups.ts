import {
  DeclarationReflection,
  ReflectionCategory,
  ReflectionGroup,
} from 'typedoc/dist/lib/models';

import { MarkdownBuilder } from '../tools/builder';
import { heading, horizontalRule } from '../tools/elements';
import { memberTemplate } from './member';

export function groupsTemplate(groups: ReflectionGroup[]) {
  const md = new MarkdownBuilder();

  groups
    ?.filter((group) => !group.allChildrenHaveOwnDocument())
    .forEach((group) => {
      if (group.categories) {
        group.categories
          ?.filter((category) => !category.allChildrenHaveOwnDocument())
          .forEach((category) => {
            md.add(heading(2, `${group.title} ${category.title}`));
            md.add(getMembers(category));
          });
      } else {
        md.add(heading(2, group.title));
        md.add(getMembers(group));
      }
    });

  return md.print();
}

function getMembers(item: ReflectionGroup | ReflectionCategory) {
  const md = new MarkdownBuilder();
  item.children.forEach((child, index) => {
    md.add(memberTemplate(child as DeclarationReflection));
    if (index !== item.children.length - 1) {
      md.add(horizontalRule());
    }
  });
  return md.print();
}
