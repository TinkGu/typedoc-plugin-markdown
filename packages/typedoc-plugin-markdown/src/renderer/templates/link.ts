import * as path from 'path';

import settings from '../settings';
import { link } from '../tools/elements';
import { escapeChars } from '../tools/utils';

export function linkTemplate(label: string, to = '', escapeLabel = true) {
  label = escapeLabel ? escapeChars(label) : label;
  return link(label, getUrl(to));
}

function getUrl(to: string) {
  const { options, activeUrl } = settings;

  const urlPrefix = /^(http|ftp)s?:\/\//;
  if (urlPrefix.test(to)) {
    return to;
  } else {
    return options.publicPath
      ? getPublicUrl(options.publicPath, to)
      : getRelativeUrl(activeUrl, to);
  }
}

function getPublicUrl(publicPath: string, url: string) {
  return publicPath + url;
}

function getRelativeUrl(from: string, to: string) {
  const relative = path.relative(path.dirname(from), path.dirname(to));
  return path.join(relative, path.basename(to)).replace(/\\/g, '/');
}
