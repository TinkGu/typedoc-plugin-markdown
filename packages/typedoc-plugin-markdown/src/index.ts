import * as path from 'path';

import {
  Application,
  Context,
  Converter,
  PageEvent,
  ParameterType,
} from 'typedoc';
import { setActiveUrl, setOptions } from './context';
import { CustomOptionsReader } from './tools/options-reader';

import { MarkdownTheme } from './theme';

export function load(app: Application) {
  addDeclarations(app);
  addSettings(app);
  setTheme(app);
}

function addDeclarations(app: Application) {
  app.options.addDeclaration({
    help: '[Markdown Plugin] Do not render page title.',
    name: 'hidePageTitle',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    help: '[Markdown Plugin] Do not render breadcrumbs in template.',
    name: 'hideBreadcrumbs',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    help: '[Markdown Plugin] Specifies the base path that all links to be served from. If omitted all urls will be relative.',
    name: 'publicPath',
    type: ParameterType.String,
  });

  app.options.addDeclaration({
    help: '[Markdown Plugin] Use HTML named anchors as fragment identifiers for engines that do not automatically assign header ids. Should be set for Bitbucket Server docs.',
    name: 'namedAnchors',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    help: '[Markdown Plugin] Output all reflections into seperate output files.',
    name: 'allReflectionsHaveOwnDocument',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    help: '[Markdown Plugin] Separator used to format filenames.',
    name: 'filenameSeparator',
    type: ParameterType.String,
    defaultValue: '.',
  });

  app.options.addDeclaration({
    help: '[Markdown Plugin] The file name of the entry document.',
    name: 'entryDocument',
    type: ParameterType.String,
    defaultValue: 'README.md',
  });

  app.options.addDeclaration({
    help: '[Markdown Plugin] Do not render in-page table of contents items.',
    name: 'hideInPageTOC',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    help: '[Markdown Plugin] Customise the index page title.',
    name: 'indexTitle',
    type: ParameterType.String,
  });
}

function addSettings(app: Application) {
  app.renderer.on(PageEvent.BEGIN, (page: PageEvent) => {
    setActiveUrl(page.url);
  });

  app.converter.on(Converter.EVENT_RESOLVE_BEGIN, (context: Context) => {
    setOptions({
      project: context.project,
      entryDocument: app.options.getValue('entryDocument') as string,
      entryPoints: app.options.getValue('entryPoints') as string[],
      readme: app.options.getValue('readme') as string,
      publicPath: app.options.getValue('publicPath') as string,
      hideBreadcrumbs: app.options.getValue('hideBreadcrumbs') as boolean,
      hideInPageTOC: app.options.getValue('hideInPageTOC') as boolean,
      namedAnchors: app.options.getValue('namedAnchors') as boolean,
    });
  });
}

function setTheme(app: Application) {
  const themeRef = app.options.getValue('theme');

  const getCustomTheme = (themeFile: string) => {
    try {
      const ThemeClass = require(themeFile);
      const instance = ThemeClass[Object.keys(ThemeClass)[0]];
      return instance.prototype instanceof MarkdownTheme ? instance : null;
    } catch (e) {
      return null;
    }
  };

  if (['default', 'markdown'].includes(themeRef)) {
    app.renderer.theme = new MarkdownTheme(app.renderer);
  } else {
    const CustomTheme = getCustomTheme(
      path.resolve(path.join(themeRef, 'theme.js')),
    );
    if (CustomTheme !== null) {
      app.options.addReader(new CustomOptionsReader());
      app.renderer.theme = new CustomTheme(app.renderer);
    } else {
      app.logger.warn(
        `[typedoc-plugin-markdown] '${themeRef}' is not a recognised markdown theme.`,
      );
    }
  }
}
