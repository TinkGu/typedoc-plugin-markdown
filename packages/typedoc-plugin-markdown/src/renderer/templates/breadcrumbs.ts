import { Reflection } from 'typedoc';
import { PageEvent } from 'typedoc/dist/lib/output/events';
import settings from '../settings';
import { MarkdownBuilder } from '../tools/builder';
import { escapeChars } from '../tools/utils';
import { linkTemplate } from './link';

export function breadcrumbsTemplate(page: PageEvent) {
  const { activeUrl, globalsName, globalsFile, options } = settings;

  const md = new MarkdownBuilder();

  md.add(
    page.url === options.entryDocument
      ? page.project.name
      : linkTemplate(page.project.name, options.entryDocument),
  );

  if (!options.readme.endsWith('none')) {
    if (activeUrl === globalsFile) {
      md.add(globalsName);
    } else {
      md.add(linkTemplate(globalsName, globalsFile));
    }
  }

  const addBreadcrumb = (model: Reflection, md: MarkdownBuilder) => {
    if (model && model.parent) {
      addBreadcrumb(model.parent, md);

      if (activeUrl === model.url) {
        md.add(escapeChars(model.name));
      } else {
        md.add(linkTemplate(model.name, model.url));
      }
    }
  };

  addBreadcrumb(page.model, md);

  return md.print(' / ');
}
