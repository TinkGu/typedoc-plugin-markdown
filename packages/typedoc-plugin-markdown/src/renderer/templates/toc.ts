import {
  DeclarationReflection,
  ProjectReflection,
  ReflectionCategory,
  ReflectionGroup,
} from 'typedoc/dist/lib/models';

import settings from '../settings';
import { MarkdownBuilder } from '../tools/builder';
import { heading, unorderedList } from '../tools/elements';
import { linkTemplate } from './link';

export function tocTemplate(model: ProjectReflection | DeclarationReflection) {
  const options = settings.options;
  const md = new MarkdownBuilder();
  const isVisible = model.groups?.some((group) =>
    group.allChildrenHaveOwnDocument(),
  );
  if ((!this.hideInPageTOC && model.groups) || (isVisible && model.groups)) {
    if (!options.hideInPageTOC) {
      md.add(heading(2, 'Table of contents'));
    }
    model.groups
      ?.filter(
        (group) => !options.hideInPageTOC || group.allChildrenHaveOwnDocument(),
      )
      .forEach((group) => {
        if (group.categories) {
          group.categories.forEach((category) => {
            md.add(getGroup(category, category.title + ' ' + group.title));
          });
        } else {
          md.add(getGroup(group, group.title));
        }
      });
  }

  return md.print();
}

function getGroup(group: ReflectionGroup | ReflectionCategory, title: string) {
  const options = settings.options;
  const md = new MarkdownBuilder();
  md.add(heading(options.hideInPageTOC ? 2 : 3, title));
  const groupItems = group.children.map((child) =>
    linkTemplate(child.name, child.url),
  );
  md.add(unorderedList(groupItems));
  return md.print();
}
