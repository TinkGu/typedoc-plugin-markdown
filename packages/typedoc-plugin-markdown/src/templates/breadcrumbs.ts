import { Reflection, PageEvent, ContainerReflection } from 'typedoc';
import { useSettings } from '../context';
import { MarkdownBuilder } from '../tools/builder';
import { escapeChars } from '../tools/utils';
import { linkTemplate } from './link';

export function breadcrumbsTemplate(page: PageEvent<ContainerReflection>) {
  const { activeUrl, globalsName, globalsFile, entryDocument, readme } =
    useSettings();

  const md = new MarkdownBuilder();

  md.add(
    page.url === entryDocument
      ? page.project.name
      : linkTemplate(page.project.name, entryDocument),
  );

  if (!readme.endsWith('none')) {
    if (activeUrl === globalsFile) {
      md.add(globalsName);
    } else {
      md.add(linkTemplate(globalsName, globalsFile));
    }
  }

  const addBreadcrumb = (model: Reflection, md: MarkdownBuilder) => {
    if (model && model.parent) {
      addBreadcrumb(model.parent, md);

      if (activeUrl === model.url) {
        md.add(escapeChars(model.name));
      } else {
        md.add(linkTemplate(model.name, model.url));
      }
    }
  };

  addBreadcrumb(page.model, md);

  return md.print(' / ');
}
