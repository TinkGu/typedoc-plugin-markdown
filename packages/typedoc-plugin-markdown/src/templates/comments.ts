import { Comment } from 'typedoc/dist/lib/models';
import { getContext } from '../context';
import { MarkdownBuilder } from '../tools/builder';
import { backTicks, bold, link } from '../tools/elements';
import { linkTemplate } from './link';

const URL_PREFIX = /^(http|ftp)s?:\/\//;
const BRACKETS = /\[\[([^\]]+)\]\]/g;
const INLINE_TAG =
  /(?:\[(.+?)\])?\{@(link|linkcode|linkplain)\s+((?:.|\n)+?)\}/gi;

export function commentsTemplate(model: Comment) {
  const md = new MarkdownBuilder();

  if (model.hasVisibleComponent()) {
    if (model.shortText) {
      md.add(comment(model.shortText));
    }
    if (model.text) {
      md.add(comment(model.text));
    }
    if (model.tags) {
      const tags = model.tags.map(
        (tag) =>
          bold(backTicks(tag.tagName)) +
          (tag.text
            ? comment((tag.text.startsWith('\n') ? '' : ' ') + tag.text)
            : ''),
      );
      md.add(tags.join('\n\n'));
    }
  }
  return md.print();
}

export function comment(text: string) {
  return replaceInlineTags(replaceBrackets(text));
}

function replaceBrackets(text: string) {
  return text.replace(BRACKETS, (match: string, content: string): string => {
    const split = splitLinkText(content);
    return buildLink(match, split.target, split.caption);
  });
}

function replaceInlineTags(text: string): string {
  return text.replace(
    INLINE_TAG,
    (match: string, leading: string, tagName: string, content: string) => {
      const split = splitLinkText(content);
      const target = split.target;
      const caption = leading || split.caption;

      return buildLink(match, target, caption, tagName === 'linkcode');
    },
  );
}

function buildLink(
  original: string,
  target: string,
  caption: string,
  monospace = false,
) {
  if (monospace) {
    caption = backTicks(caption);
  }

  if (URL_PREFIX.test(target)) {
    return link(caption, target);
  }

  const { project } = getContext();

  const reflection = project.findReflectionByName(target);
  if (reflection && reflection.url) {
    return linkTemplate(caption, reflection.url, !monospace);
  } else {
    return original;
  }
}

function splitLinkText(text: string) {
  let splitIndex = text.indexOf('|');
  if (splitIndex === -1) {
    splitIndex = text.search(/\s/);
  }
  if (splitIndex !== -1) {
    return {
      caption: text
        .substr(splitIndex + 1)
        .replace(/\n+/, ' ')
        .trim(),
      target: text.substr(0, splitIndex).trim(),
    };
  } else {
    return {
      caption: text,
      target: text,
    };
  }
}
