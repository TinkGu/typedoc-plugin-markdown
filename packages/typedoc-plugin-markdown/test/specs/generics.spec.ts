import { PageEvent } from 'typedoc/dist/lib/output/events';
import * as context from '../../src/context';
import * as breadcrumbsTemplate from '../../src/templates/breadcrumbs';
import * as commentsTemplate from '../../src/templates/comments';
import { declarationTemplate } from '../../src/templates/declaration';
import * as groupsTemplate from '../../src/templates/groups';
import * as linkTemplate from '../../src/templates/link';
import { pageTemplate } from '../../src/templates/page';
import { signatureTemplate } from '../../src/templates/signature';
import * as sourcesTemplate from '../../src/templates/sources';
import * as tocTemplate from '../../src/templates/toc';
import { formatContents } from '../../src/tools/utils';
import { TestApp } from '../test-app';

describe(`Generics:`, () => {
  let testApp: TestApp;

  beforeAll(() => {
    jest.spyOn(groupsTemplate, 'groupsTemplate').mockReturnValue('[GROUPS]');
    jest
      .spyOn(breadcrumbsTemplate, 'breadcrumbsTemplate')
      .mockReturnValue('[BREADCRUMBS]');
    jest
      .spyOn(commentsTemplate, 'commentsTemplate')
      .mockReturnValue('[COMMENT]');
    jest.spyOn(tocTemplate, 'tocTemplate').mockReturnValue('[TOC]');
    jest.spyOn(linkTemplate, 'linkTemplate').mockReturnValue('[LINK]');
    jest.spyOn(sourcesTemplate, 'sourcesTemplate').mockReturnValue('[SOURCES]');

    jest
      .spyOn(context, 'getContext')
      .mockReturnValue({ activeUrl: 'modules/generics.md' } as any);

    testApp = new TestApp(['generics.ts']);
    testApp.bootstrap();
  });

  afterAll(() => {
    jest.restoreAllMocks();
    testApp.cleanup();
  });

  test(`should compile class with type params`, () => {
    expect(
      formatContents(
        pageTemplate({
          url: 'ClassWithTypeParams.md',
          model: testApp.findReflection('ClassWithTypeParams'),
          project: testApp.project,
        } as PageEvent),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile function with a simple type param'`, () => {
    expect(
      formatContents(
        signatureTemplate(
          testApp.findReflection('functionWithTypeParam').signatures[0] as any,
        ),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile function with complex type params'`, () => {
    expect(
      formatContents(
        signatureTemplate(
          testApp.findReflection('functionWithTypeParams').signatures[0] as any,
        ),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile type with nested generics'`, () => {
    expect(
      formatContents(
        declarationTemplate(testApp.findReflection('nestedGenerics')),
      ),
    ).toMatchSnapshot();
  });
});
