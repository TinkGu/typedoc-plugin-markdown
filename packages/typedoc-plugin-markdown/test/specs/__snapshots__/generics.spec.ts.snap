// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generics: should compile class with type params 1`] = `
"[BREADCRUMBS]

# Class: ClassWithTypeParams<T, V\\\\>

[COMMENT]

## Type parameters

| Name | Description |
| :------ | :------ |
| \`T\` | [COMMENT] |
| \`V\` | [COMMENT] |

[TOC]

[GROUPS]
"
`;

exports[`Generics: should compile function with a simple type param' 1`] = `
"▸ \`Const\` **functionWithTypeParam**<\`A\`\\\\>(): \`boolean\`

[COMMENT]

#### Type parameters

| Name |
| :------ |
| \`A\` |

#### Returns

\`boolean\`

[SOURCES]
"
`;

exports[`Generics: should compile function with complex type params' 1`] = `
"▸ \`Const\` **functionWithTypeParams**<\`A\`, \`B\`, \`C\`\\\\>(): \`boolean\`

[COMMENT]

#### Type parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| \`A\` | extends [LINK]<\`string\`, \`number\`, \`A\`\\\\> | [COMMENT] |
| \`B\` | \`string\` \\\\| \`boolean\` | [COMMENT] |
| \`C\` | \`string\` | - |

#### Returns

\`boolean\`

[SOURCES]
"
`;

exports[`Generics: should compile type with nested generics' 1`] = `
"Ƭ **nestedGenerics**: [LINK]<[LINK]<[LINK]<\`string\`\\\\>\\\\>\\\\>

[COMMENT]

[SOURCES]
"
`;
